### DISCUSSION TITLE
Unexpected behaviour with SLOPPINESS and BASEDIR

### DISCUSSION THREAD

**Discussion by fjavierv** [ccache/ccache#332] [Support]
After read documents and closed issues, I have some doubts still about ccache. I will try to expose them with two scenarios (using relative path and using absolute path).

We have two folder "a" and "b" with the same simple "main.cpp" inside. The "main.cpp" file has the following content:

```
// main.cpp file
int fname() {}
```
We are using ccache 3.4.2 and this is our configuration:

```
max_size = 20000.0M
hard_link = false
umask = 002
hash_dir = false
```
# First scenario (using relative path):
We start cleaning the cache:

`$> ccache -C -z`

We compile the "main.cpp" in the "a" folder using relative path:

`$> ccache g++ -c a/main.cpp`

As might be expected I got "cache miss". Now we compile "main.cpp" in the "b" folder

`$> ccache g++ -c b/main.cpp`

I got "cache miss" again. Now I compile "main.cpp" inside of "a" folder:

`$> cd a && ccache g++ -c main.cpp`

I got "cache miss" again. After that I defined CCACHE_SLOPINNES as follow:

```
$> export CCACHE_SLOPPINESS=file_macro,time_macros`
$> cd ..
```

And I repeat the test, so firt step is to clean cache:

`$> ccache -C -z `

We compile the "main.cpp" in the "a" folder using relative path:

`$> ccache g++ -c a/main.cpp`

I get "cache miss". Now we compile "main.cpp" in the "b" folder

`$> ccache g++ -c b/main.cpp`

Surprisingly I got "cache hit direct" I don't understand why, but we following compiling "main.cpp" from "a" folder:

`$> cd a && ccache g++ -c main.cpp`

We got again "cache hit direct" again and I didn't know why.

Summary:
```
$> ccache -C -z
$> ccache g++ -c a/main.cpp --> cache miss
$> ccache g++ -c b/main.cpp --> cache miss
$> cd a && ccache g++ -c main.cpp --> cache miss
$> cd ..
$> export CCACHE_SLOPPINESS=file_macro,time_macros
$> ccache -C -z 
$> ccache g++ -c a/main.cpp --> cache miss
$> ccache g++ -c b/main.cpp --> cache direct hit
$> cd a && ccache g++ -c main.cpp --> cache direct hit
```

# Second scenario (using absolute path):
We start cleaning the cache and unsetting CCACHE_SLOPPINESS:
```
$> ccache -C -z && unset CCACHE_SLOPPINESS
$> ccache g++ -c /home/username1/Tests/a/main.cpp --> cache miss
$> ccache g++ -c /home/username1/Tests/b/main.cpp --> cache miss
$> ccache g++ -c /home/username2/Tests/a/main.cpp --> cache miss
$> ccache g++ -c /home/username2/Tests/b/main.cpp --> cache miss
```

We clean cache again and set basedir variable:
```
$> ccache -C -z
$> export CCACHE_BASEDIR=/home
$> ccache g++ -c /home/username1/Tests/a/main.cpp --> cache miss
$> ccache g++ -c /home/username1/Tests/b/main.cpp --> cache miss
$> ccache g++ -c /home/username2/Tests/a/main.cpp --> cache direct hit
$> ccache g++ -c /home/username2/Tests/b/main.cpp --> cache direct hit
```

We clean cache again and set sloppinnes variable also:
```
$> ccache -C -z
$> export CCACHE_SLOPPINESS=file_macro,time_macros
$> ccache g++ -c /home/username1/Tests/a/main.cpp --> cache miss
$> ccache g++ -c /home/username1/Tests/b/main.cpp --> cache direct hit
$> ccache g++ -c /home/username2/Tests/a/main.cpp --> cache direct hit
$> ccache g++ -c /home/username2/Tests/b/main.cpp --> cache direct hit
```

Maybe this is the expected behaviour but It seem strange.

**Comment by afbjorklund** [comment]
If you set direct `sloppiness` to include `SLOPPY_FILE_MACRO`, then the file name will **not** be hashed.
For your second test, I cannot reproduce the before values - get cache misses, for different directories.

**Comment by jrosdahl** [comment]
I made some documentation improvements in 50719ef3a71d8afc669f2192271e7e0cceffe045. @fjavierv: Does that make things more clear or is anything still unexpected?
